File: TypeScript/day01/day01.ts
import * as fs from "fs";
import * as path from "path";

const pathOfFile = path.join(__dirname, "../../challenges/day01/input.txt");
const bothLists = readFile();
const santasFirstList = bothLists[0];
const santasSecondList = bothLists[1];

function readFile() {
  const fileContent = fs.readFileSync(pathOfFile, "utf-8");
  const lines = fileContent.split("\n");
  const leftList: Array<number> = [];
  const rightList: Array<number> = [];

  for (let line of lines) {
    const parts = line.split("   ");
    const number1 = parseInt(parts[0]);
    const number2 = parseInt(parts[1]);
    leftList.push(number1);
    rightList.push(number2);
  }

  return [leftList, rightList];
}
//Part 1
function calculateDifferencePart1(
  santasFirstList: Array<number>,
  santasSecondList: Array<number>
) {
  const sortedFirstList = santasFirstList.sort((a, b) => a - b);
  const sortedSecondList = santasSecondList.sort((a, b) => a - b);

  let totalDifference = 0;

  for (let i = 0; i < sortedFirstList.length; i++) {
    const firstNumber = sortedFirstList[i];
    const secondNumber = sortedSecondList[i];
    if (firstNumber > secondNumber) {
      totalDifference += firstNumber - secondNumber;
    } else if (secondNumber > firstNumber) {
      totalDifference += secondNumber - firstNumber;
    }
  }

  return totalDifference;
}

const part1Difference = calculateDifferencePart1(
  santasFirstList,
  santasSecondList
);

//Part 2
function countSimilar(
  santasFirstList: Array<number>,
  santasSecondList: Array<number>
) {
  let score = 0;

  for (let i = 0; i < santasFirstList.length; i++) {
    const firstNumber = santasFirstList[i];
    const sameNumbersInSecondList = santasSecondList.filter(
      (number) => number === firstNumber
    ).length;

    score += firstNumber * sameNumbersInSecondList;
  }
  return score;
}

const part2Score = countSimilar(santasFirstList, santasSecondList);
console.log(`Solution part1: ${part1Difference}`);
console.log(`Solution part2: ${part2Score}`);
//Part 1 solution: 3246517
//Part 2 solution: 29379307

File: Python/day01/day01.py
import os

current_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(current_dir, "../../challenges/day01/input.txt")

with open(file_path, "r") as file:
    lines = file.readlines()

left_list = []
right_list = []

for line in lines:
    formated_line = line.strip().split("   ")
    left_list.append(int(formated_line[0]))
    right_list.append(int(formated_line[1]))

def get_difference_part1(left_list, right_list): 
    sum = 0
    left_list.sort()
    right_list.sort()
    for i in range(len(left_list)):
        if left_list[i] > right_list[i]:
            sum += left_list[i] - right_list[i]
        else: 
            sum += right_list[i] - left_list[i]
    return sum

print(get_difference_part1(left_list, right_list))

def get_similarity_part2(left_list, right_list):
    sum = 0
    for i in range(len(left_list)):
        if left_list[i] in right_list:
            sum += left_list[i] * right_list.count(left_list[i])
    return sum

print(get_similarity_part2(left_list, right_list))